File: app/src/main/java/com/example/jacob/myapplication/ShareScreenshotTask.java
package com.example.jacob.myapplication;

import android.app.Activity;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;
import android.util.Log;
import android.view.View;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.Date;

/**
 * Created by jacob on 24/11/2015.
 */
public class ShareScreenshotTask extends AsyncTask<Void, Void, String>{

    private Activity myAct;
    public ShareScreenshotTask(Activity myAct) {
        this.myAct = myAct;
    }

    @Override
    protected String doInBackground(Void... params) {
        return saveBitmap();
    }

    @Override
    protected void onPostExecute(String aVoid) {
        super.onPostExecute(aVoid);
        sendIntent(aVoid);
    }

    public String saveBitmap() {
        Bitmap myBitmap;
        View v1 = myAct.findViewById(android.R.id.content); //this works too
        v1.setDrawingCacheEnabled(true);
        myBitmap = v1.getDrawingCache();
        Date now = new Date();
        android.text.format.DateFormat.format("yyyy-MM-dd_hh:mm:ss", now);
        File dir = new File(Environment.getExternalStorageDirectory()
                + File.separator + "WhatsAppAnalyzer");
        if (!dir.exists())
            dir.mkdir();
        String filePath = Environment.getExternalStorageDirectory()
                + File.separator + "WhatsAppAnalyzer" + File.separator + now +  ".png";
        File imagePath = new File(filePath);
        FileOutputStream fos;
        try {
            fos = new FileOutputStream(imagePath);
            myBitmap.compress(Bitmap.CompressFormat.PNG, 100, fos);
            fos.flush();
            fos.close();
        } catch (FileNotFoundException e) {
            Log.e("GREC", e.getMessage(), e);
        } catch (IOException e) {
            Log.e("GREC", e.getMessage(), e);
        }
        v1.setDrawingCacheEnabled(false);
        return filePath;
    }

    public void sendIntent(String path) {
        Intent intent = new Intent(android.content.Intent.ACTION_SEND);
        intent.putExtra(android.content.Intent.EXTRA_TEXT,
                "Generated with Whatsapp Analyzer");
        intent.setType("image/png");
        Uri myUri = Uri.parse("file://" + path);
        intent.putExtra(Intent.EXTRA_STREAM, myUri);
        myAct.startActivity(Intent.createChooser(intent, "Send image"));
    }
}

File: app/src/main/java/me/jacobrr/whatsappanalyzer/tasks/ShareScreenshotTask.kt
package me.jacobrr.whatsappanalyzer.tasks

import android.app.Activity
import android.content.ContentValues
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.AsyncTask
import android.provider.MediaStore
import android.util.Log
import android.view.View
import androidx.core.view.drawToBitmap
import java.io.FileNotFoundException
import java.io.IOException
import java.util.*

/**
 * Created by jacob on 24/11/2015.
 */
class ShareScreenshotTask(private val myAct: Activity) : AsyncTask<Void, Void, Uri>() {

    override fun doInBackground(vararg params: Void): Uri {
        return saveBitmap()
    }

    override fun onPostExecute(aVoid: Uri) {
        super.onPostExecute(aVoid)
        sendIntent(aVoid)
    }

    private fun saveBitmap(): Uri {
        val v1 = myAct.findViewById<View>(android.R.id.content) //this works too
        val myBitmap = v1.drawToBitmap()
        val now = Date()
        val name = android.text.format.DateFormat.format("yyyy-MM-dd_hh:mm:ss", now)

        val resolver = myAct.applicationContext.contentResolver

        val imageCollection = MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL)
        val imageData = ContentValues().apply {
            put(MediaStore.Images.Media.DISPLAY_NAME, "$name")
            put(MediaStore.Images.Media.MIME_TYPE, "image/png")
            put(MediaStore.Images.Media.RELATIVE_PATH, "Pictures/Screenshots/")
            put(MediaStore.Images.Media.IS_PENDING, 1)
        }
        val uri = resolver.insert(imageCollection, imageData)

        try {
            resolver.openOutputStream(uri!!).use { out ->
                myBitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
            }

            imageData.clear()
            imageData.put(MediaStore.Images.Media.IS_PENDING, 0)
            resolver.update(uri, imageData, null, null)
        } catch (e: FileNotFoundException) {
            Log.e("GREC", e.message, e)
        } catch (e: IOException) {
            Log.e("GREC", e.message, e)
        }
        return uri!!
    }

    fun sendIntent(path: Uri) {
        val intent = Intent(Intent.ACTION_SEND)
        intent.putExtra(Intent.EXTRA_TEXT,
                "Generated with Whatsapp Analyzer")
        intent.type = "image/png"
        intent.putExtra(Intent.EXTRA_STREAM, path)
        myAct.startActivity(Intent.createChooser(intent, "Send image"))
    }
}
